### ISSUE TITLE
tf.io.gfile.listdir is inconsistent between GCS dir and local dir - adds trailing slashes

### ISSUE THREAD

**Issue by Ark-kun** [tensorflow/tensorflow#39167]
**System information**
- TensorFlow version (use command below): v2.0.0-rc2-26-g64c3d38 2.0.0
- Python version: 3.6

**Describe the current behavior**

When listing the directory items for a GCS dir, listdir returns names with trailing slashes. 
When listing the directory items for a local dir, listdir returns names without trailing slashes. 

**Describe the expected behavior**

The behavior needs to be consistent.

**Standalone code to reproduce the issue**

```
tensorflow.io.gfile.listdir('gs://bucket/dir')
>>> ['eval/', 'train/']

tensorflow.io.gfile.listdir('..')
>>> ['eval', 'train']
```

**Comment by mihaimaruseac** [comment]
Can you please test if this also happens on 2.2-rc4? On tf-nightly?

**Comment by Ark-kun** [comment]
> Can you please test if this also happens on 2.2-rc4?

I've tested this on 2.2.0. It's still broken.


```
tensorflow.io.gfile.listdir('gs://gcp-public-data-landsat')

...
 'LC08/',
 'LE00/',
 'LE07/',
 'LM01/',
 'LM02/',
 'LM03/',
 'LM04/',
 'LM05/',
 'LO08/',
 'LT00/',
 'LT04/',
 'LT05/',
 'LT08/']
```

```
tensorflow.io.gfile.listdir('..')

['sys',
 'opt',
 'boot',
 'bin',
 'home',
 'usr',
...
```

**Comment by mihaimaruseac** [comment]
Does still reproduce in tf-nightly? I recall sending a change that touched the `/` at the end of the paths.

**Comment by Saduf2019** [comment]
@Ark-kun 
Please update as per above comment, If this is still an issue in nightly.

**Comment by mihaimaruseac** [comment]
This should be fixed in the 2.4 release, after [rewriting implementation](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/platform/file_system_helper.cc;l=123-266;drc=8b5b9dc96666a3a5d27fad7179ff215e3b74b67c)

**Comment by 0x0badc0de** [comment]
Still the case with TF 2.4.1.
Local dir - no trailing slashes
s3 bucket - no trailing slashes
gcs bucket - WITH trailing slashes

**Comment by 0x0badc0de** [comment]
Can it be protected from auto-closing since it's clearly not fixed and easy to reproduce?

**Comment by NA-Dev** [comment]
I am facing this issue now as well on 2.4. Definitely not resolved. I will not be following this every month to make sure it doesn't go  stale, but wanted to let someone know.

**Comment by gmarchetti2020** [comment]
still the case in 2.5

**Comment by tilakrayal** [comment]
@Ark-kun ,
Is this still an issue? Could you please confirm in TensorFlow latest stable version v.2.7. Thanks!

**Comment by Ark-kun** [comment]
> @Ark-kun , Is this still an issue? Could you please confirm in TensorFlow latest stable version v.2.7. Thanks!

Yes, it's still an issue in 2.7.0. I'm not sure this issue will fix itself without an explicit fix.

/reopen

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by mihaimaruseac** [comment]
This is still an issue.

**Comment by Venkat6871** [comment]
Hi @Ark-kun ,
Thank you so much for filing this bug request! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.To avoid duplicate work, please leave a comment here if you decide to take it on!
Thank you again for your interest and we look forward to seeing your contributions!
